libexec/rtld-elf/aarch64/rtld_machdep.h

```diff
@@ -69,13 +69,25 @@ uintptr_t reloc_jmpslot(uintptr_t *where, uintptr_t target,
 
 #if __has_feature(capabilities)
 
-#define	FUNC_PTR_REMOVE_PERMS						\
+#define	REMOVE_CODE_SW_PERMS						\
+	(CHERI_PERMS_SWALL & ~CHERI_PERM_SYSCALL)
```

Comment:
Note that currently `CHERI_PERM_SYSCALL` doesn't do anything.  It only ever did for libcheri on MIPS I think.

Is one consequence of this that we can't permit use of software permissions to mean anything in userspace now?  I wonder how we feel about that.  I kind of feel like the kernel should gate access to SW perms that have meaning in the kernel, and similar here.  That is, I don't know that default-to-deny is the right policy for rtld for SW perms that are currently unused?  If we do want to default-to-deny then probably the kernel should be using default-to-deny.  For either of those, if rtld and kernel are in agreement on the default, then rtld should really only be stripping explicit permissions with known semantics that it has received from the kernel but doesn't want to delegate.

Reply:
Also, is "Fix aa64" a typo in the log meant to reference the issue opened about this?

Reply:
I think it was a leftover from a git squash.

Reply:
`CHERI_PERM_SYSCALL` is enforced because I have been getting syscall failures due to this. If we want to remove it, it is a different question, but I think it is definitely used even if it is a leftover perhaps.

I think this patch had two separate purposes, that probably can be split up if necessary:
1. We need to strip the SW_VMEM permission bit from the capability used to do shared library relocations that we get from mmap when we load objects in memory.
2.  I agree that the kernel should gate the ones that have meaning in the kernel, which I think is being done. Rtld is stripping the permission with known semantic. However point (1) and the question from Kevin #1656 made me question what to do about the SW bits of pointers that are produced as a result of relocations. If the user program gives them a specific meaning, does it make sense that pointers in the GOT have that bit? Probably stripping it by default is as bad as leaving it by default because it depends on the semantic of the SW bit, however this probably means that the user code should have a way to control this part of the relocation mechanism if SW bits are going to be used.

Reply:
Currently SW bits can be partially used by the program as they are present in the main() PCC, however I am unsure about this choice because to do this properly rtld should give out a pair of code and data capabilities with SW bits that can be used within the program to produce any other capability where the SW bit has a specific user-defined semantic. Although I can't entirely wrap my mind about how this would look like.

Reply:
I wonder if ideally you'd want an ELF note or flag to tell the kernel you wanted a particular user permission by default an otherwise you'd request it explicitly via new mmap PROT_ values? Not really sure. The big thing is probably that if everything defaults to having a permission then it doesn't have any meaning so you need the ability to restrict creation which implies restricting mmap and shmat.

Reply:
The ELF note/flag is probably the correct approach. I think I will revert the clearing of SW bits in rtld and only strip SW_VMEM (see point (1) above).